src/application/transaction/verifyTx.ts

```diff
@@ -0,0 +1,302 @@
+    assertResource(action.rri, to.resource),
+    isSelfTransfer ? ok(null) : assertAmount(action.amount, to.amount),
+  ])
+}
+
+const verifyStake = (
+  action: IntendedStakeTokensAction,
+  instructions: PreparedStakeGroup,
+) => {
+  const [from, to] = instructions
+  return combineWithAllErrors([
+    assertNumberOfInstructions(2, instructions.length),
+    assertSubstateTypes(
+      [SubStateType.TOKENS, SubStateType.PREPARED_STAKE],
+      instructions.map(({ substateType }) => substateType),
+    ),
+    assertAccountAddress('from', action.from_account, from.owner),
+    assertAccountAddress('to', action.to_validator, to.validator),
+    assertResource(action.rri, from.resource),
+    assertAmount(action.amount, to.amount),
+  ])
+}
+
+const verifyUnstake = (
+  action: IntendedUnstakeTokensAction,
+  instructions: PreparedUnstakeGroup,
+) => {
+  const [from, to] = instructions
+  return combineWithAllErrors([
+    assertNumberOfInstructions(2, instructions.length),
+    assertSubstateTypes(
+      [SubStateType.PREPARED_UNSTAKE, SubStateType.STAKE_OWNERSHIP],
+      instructions.map(({ substateType }) => substateType),
+    ),
+    assertAccountAddress('from', action.from_validator, from.validator),
+    assertAccountAddress('to', action.to_account, to.owner),
+  ])
+}
+
+const verifyTxIntent =
+  (txIntent: TransactionIntent) => (groups: InstructionGroup[]) => {
+    const [, ...instructionGroups] = groups
+
+    return combine(
+      txIntent.actions
+        .map((action, index) => {
+          const currentGroup = instructionGroups[index]
+
+          switch (action.type) {
+            case ActionType.TRANSFER:
+              return verifyTransfer(action, currentGroup as TokensGroup)
+
+            case ActionType.STAKE:
+              return verifyStake(action, currentGroup as PreparedStakeGroup)
+
+            case ActionType.UNSTAKE:
+              return verifyUnstake(action, currentGroup as PreparedUnstakeGroup)
+
+            default:
+              return ok('')
+          }
+        })
+        .flatMap(item => item),
+    )
+  }
+
+export const verifyTx =
+  (txIntent: TransactionIntent, network: Network, trackError: TrackError) =>
+  (tx: BuiltTransaction) => {
+    return Transaction.fromBuffer(Buffer.from(tx.transaction.blob, 'hex'))
+      .map(tx => {
+        log.info(tx.toString())
+        return tx.instructions
+      })
+      .map(filterUpAndEndInstructions)
+      .map(groupInstructionsByActions)
+      .map(parseGroupData(network))
+      .andThen(group => combine(group))
+      .map(groups =>
+        groups.filter((group): group is InstructionGroup => group.length > 0),
+      )
+      .andThen(verifyTxIntent(txIntent))
+      .map(() => tx)
+      .mapErr(errors => {
+        trackError({
+          errors: [
+            radixAPIError({
+              message: 'transaction intent does not match built transaction',
+              details: { type: 'VerifyBuiltTransactionError', errors },
```

Comment:
The `errors` variable here can have several types. Mostly problematic might be the fact that we can expect `string[]` but also `Error`.

We probably want to change the `string` error types to `Error` to be consistent with the return type of `Transaction.fromBuffer`, and then properly display the messages using something like `errors.map(e => e.message)`.

Reply:
Good idea. Changed string to Error